### ISSUE TITLE
Easier lock interface?  Unlock on push?

### ISSUE THREAD

**Issue by theadamrippon** [ForkIssues/TrackerWin#1862]
I'm trying to figure out a decent workflow for Unreal Engine development with a team of not-particularly-technical artists (who all specifically detest Perforce).  They're all used to Fork from other game projects that didn't rely on an engine that used so many binary asset files.  I attempted to switch them over to using the UE Git plugin, but they're really struggling with it.

To use Fork for UE, they need to be able to see all the files that are in the repo, rather than just the ones that are changed, which doesn't seem to be easy right now - is there currently a way to do this?  To get around it, they tried locking via UE and then pushing from Fork, but that left all the locked files locked after push.  If nothing else, would it be possible to add a "unlock pushed files" checkbox if some of the files being pushed are currently locked?

**Comment by DanPristupov** [comment]
> see all the files that are in the repo, rather than just the ones that are changed, which doesn't seem to be easy right now - is there currently a way to do this?

The FileTree tab shows all the repository files.

<img width="525" alt="Screenshot 2023-04-11 at 9 05 30" src="https://user-images.githubusercontent.com/618115/231082385-236c0219-9380-41c7-a8e4-de49c49b8373.png">

> To get around it, they tried locking via UE and then pushing from Fork, but that left all the locked files locked after push

Git doesn't have locks. Do you mean LFS lock?

> If nothing else, would it be possible to add a "unlock pushed files" checkbox if some of the files being pushed are currently locked?

I think that is technically impossible. We push particular branches and we don't know what changes they contain.

**Comment by theadamrippon** [comment]
It looks like that file tree view is just for things that have already been committed?

Yeah, I am referring to LFS locks.  In the UE editor, when you submit files it performs the push, and then every file that was pushed is LFS unlocked after the push is performed successfully.  It's pretty straightforward implementation there, but my art team would prefer to use Fork for everything if they can.  But currently manually unlocking files after they push them is a lot to ask of artists.

**Comment by DanPristupov** [comment]
> It looks like that file tree view is just for things that have already been committed?

That's right. It's the file tree of the repository at the commit. I think you can't LFS-track (and lock) uncommitted files anyway.

> But currently manually unlocking files after they push them is a lot to ask of artists.

There is `Repository -> GitLFS -> Locks` view which shows the active locks. It is still manual unlocking, but it shows the person who set the lock.

<img width="508" alt="Screenshot 2023-04-11 at 19 30 17" src="https://user-images.githubusercontent.com/618115/231242600-dfaaeeec-9299-4db3-8110-1b1ee584a873.png">

**Comment by theadamrippon** [comment]
What about an unlock all button on that screen?  Would that be doable?  Usually they're doing dozens or hundreds of them at a time.

**Comment by DanPristupov** [comment]
We can implement ability to unlock multiple files in the next update.

However I'm not sure about the 'Update all' button. To be honest, I'm not a big fan of features like 'Stage All and Commit' or 'Commit and Push' as they are error prone.

**Comment by Sulfur-CK** [comment]
@DanPristupov I understand the reluctance. As a programmer I have no issue with staging and then committing. But content creators (tech-design/tech-art/level-design/general-design) get really confused by these concepts especially if they used Perforce previously. They prefer github desktop due to it's simpler interface, which is not great because it does not have all the nice features of Fork (the view it presents is closer to that of Perforce).

Similarly, if there is an `unlock all non-locally modified files` button, it would be a huge relief for our team. At the moment, due to the long-standing lfs issue of unlocking after push (https://github.com/git-lfs/git-lfs/issues/2671) our team has a lot of friction with locks. I am trying to convince them to use the `Repository -> Git LFS -> Status Locks` (which has been quite improved, thank you for that!) but it's difficult.

They are (rightfully) looking for an automatic way for the files to get unlocked if their PR is merged. The compromise they are willing to accept is a button that does something along the lines of "unlock all files that are currently not locally modified but are locked". Considering that Fork can already populate the list of locked files by username, I am really hoping Fork can implement something like this.

**Comment by CasperTheCat** [comment]
We've hit a similar set of issues regarding automatic LFS unlock, and I thought I'd give our interim solution to the issue. Having UI integration would be awesome for non-technically inclined users.
I've looked into this, and as far as I can tell, there is an acceptable way to automatically unlock LFS locks. LFS can't do it directly as git lacks a post-push hook, but we can shop the method they considered.

My solution was to get all unpushed commits before issuing a push, gather all modified files in those commits, doing a union against the user-held locks, and release the locks if the push succeeded.

For completeness, I was getting unpushed commits with 'git cherry remote/branch' prior to push and the modified files with 'git diff-tree --no-commit-id --name-only commitId'.
